File: src/main/java/uk/co/lukestevens/jdbc/ConfiguredDatabase.java
package uk.co.lukestevens.jdbc;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;

import javax.inject.Inject;

import uk.co.lukestevens.annotations.SetupConfig;
import uk.co.lukestevens.config.Config;

/**
 * A database implementation using {@link AppConfig}
 * properties to connect to the database.
 * 
 * @author luke.stevens
 */
public class ConfiguredDatabase extends AbstractDatabase {

	private final Config config;
	
	/**
	 * Creates a new configured database from a set of 
	 * AppConfig properties
	 * @param config The Config instance. The following properties must be specified:
	 * <ul>
	 * 	<li><code>database.url</code></li>
	 * 	<li><code>database.username</code></li>
	 * 	<li><code>database.password</code></li>
	 * </ul>
	 * 
	 */
	@Inject
	public ConfiguredDatabase(@SetupConfig Config config) {
		this.config = config;
	}

	@Override
	protected Connection getConnection() throws SQLException {
		return DriverManager.getConnection(
				config.getAsString("database.url"), 
				config.getAsString("database.username"), 
				config.getAsString("database.password"));
	}



}

File: src/main/kotlin/uk/co/lukestevens/jdbc/ConfiguredDatabase.kt
package uk.co.lukestevens.jdbc

import uk.co.lukestevens.jdbc.result.DatabaseResult
import uk.co.lukestevens.jdbc.result.WrappedDatabaseResult
import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.SQLException
import java.sql.Statement
import java.sql.Types
import java.util.*
import javax.sql.DataSource

/**
 * A further implementation of the Database interface
 * that only leaves the connection to be defined by
 * overriding classes.
 *
 * @author luke.stevens
 */
class ConfiguredDatabase(private val dataSource: DataSource) : Database {
    /**
     * Prepare a parameterised statement from a raw SQL
     * query and parameters
     * @param conn The database connection this statement will be
     * used for.
     * @param query The SQL query to prepare
     * @param params A variable array of parameters to substitute
     * into the SQL query
     * @return A prepared statement that can be used to query the database
     * @throws SQLException If a database error occurs or this method is
     * called with a closed connection
     */
    fun prepareStatement(conn: Connection, query: String, vararg params: Any): PreparedStatement {
        return conn.prepareStatement(query, Statement.RETURN_GENERATED_KEYS).apply {
            for (i in params.indices) {
                var o = params[i]
                if (o.javaClass.isEnum) {
                    o = o.toString()
                }
                if (o is Date) {
                    setObject(i + 1, o, Types.TIMESTAMP)
                } else {
                    setObject(i + 1, o)
                }
            }
        }
    }

    override fun query(query: String, vararg params: Any): DatabaseResult {
        val conn = dataSource.connection
        val rs = prepareStatement(conn, query, *params).executeQuery()
        return WrappedDatabaseResult(conn, rs)
    }

    override fun update(query: String, vararg params: Any): Long? {
        return dataSource.connection.use { conn ->
            val stmt = prepareStatement(conn, query, *params)
            stmt.executeUpdate()
            stmt.generatedKeys.let {
                if(it.next()) it.getLong(1)
                else null
            }
        }
    }
}